src/regional_settings.cpp

```diff
@@ -276,98 +363,58 @@ static void load_overmap_feature_flag_settings( const JsonObject &jo,
-                                    overmap_lake_settings.lake_depth, !overlay );
-
-        if( !overmap_lake_settings_jo.has_array( "shore_extendable_overmap_terrain" ) ) {
-            if( !overlay ) {
-                overmap_lake_settings_jo.throw_error( "shore_extendable_overmap_terrain required" );
-            }
-        } else {
-            const std::vector<std::string> from_json =
-                overmap_lake_settings_jo.get_string_array( "shore_extendable_overmap_terrain" );
-            overmap_lake_settings.unfinalized_shore_extendable_overmap_terrain.insert(
-                overmap_lake_settings.unfinalized_shore_extendable_overmap_terrain.end(), from_json.begin(),
-                from_json.end() );
-        }
-
-        if( !overmap_lake_settings_jo.has_array( "shore_extendable_overmap_terrain_aliases" ) ) {
-            if( !overlay ) {
-                overmap_lake_settings_jo.throw_error( "shore_extendable_overmap_terrain_aliases required" );
-            }
+    return om_lake_factory.is_valid( *this );
+}
+void om_settings_lake::load_om_settings_lake( const JsonObject &jo, const std::string &src )
+{
+    om_lake_factory.load( jo, src );
+}
+void om_settings_lake::load( const JsonObject &jo, const std::string_view )
+{
+    // if these settings aren't loaded, we just revert to the defaults.
+    optional( jo, was_loaded, "noise_threshold_lake", noise_threshold_lake );
+    optional( jo, was_loaded, "lake_size_min", lake_size_min );
+    optional( jo, was_loaded, "lake_depth", lake_depth );
+    mandatory( jo, was_loaded, "shore_extendable_overmap_terrain", shore_extendable_overmap_terrain );
+    if( !jo.has_array( "shore_extendable_overmap_terrain_aliases" ) ) {
+                jo.throw_error( "shore_extendable_overmap_terrain_aliases required" );
         } else {
-            for( JsonObject alias_entry :
-                 overmap_lake_settings_jo.get_array( "shore_extendable_overmap_terrain_aliases" ) ) {
-                shore_extendable_overmap_terrain_alias alias;
-                alias_entry.read( "om_terrain", alias.overmap_terrain );
-                alias_entry.read( "alias", alias.alias );
-                alias.match_type = alias_entry.get_enum_value<ot_match_type>( "om_terrain_match_type",
-                                   ot_match_type::contains );
-                overmap_lake_settings.shore_extendable_overmap_terrain_aliases.emplace_back( alias );
-            }
+        for( JsonObject alias_entry : jo.get_array( "shore_extendable_overmap_terrain_aliases" ) ) {
+            shore_extendable_overmap_terrain_alias alias;
+            alias_entry.read( "om_terrain", alias.overmap_terrain );
+            alias_entry.read( "alias", alias.alias );
+            alias.match_type = alias_entry.get_enum_value<ot_match_type>( "om_terrain_match_type",
+                               ot_match_type::contains );
+            shore_extendable_overmap_terrain_aliases.emplace_back( alias );
         }
```

Comment:
This can go back to the version before https://github.com/CleverRaven/Cataclysm-DDA/pull/70108/commits/f98e5c1fcb0c33299d12266026675a5aeceefebc if you add a deserialize function to `shore_extendable_overmap_terrain_alias` that does what this is doing.
e.g.
```cpp
void  shore_extendable_overmap_terrain_alias::deserialize(const JsonObject &jo) {
    jo.read( "om_terrain", overmap_terrain );        
    jo.read( "alias", alias );
    jo.match_type = jo.get_enum_value<ot_match_type>( "om_terrain_match_type", ot_match_type::contains );
}
```

<details>

<summary>Going through the compiler error about it:</summary>

This section tells you where the error is occurring.
```
Error: src/flexbuffer_json-inl.h:383:20: error: no matching member function for call to 'read'
            if( jv.read( element, throw_on_error ) ) {
                ~~~^~~~
src/flexbuffer_json-inl.h:1065:28: note: in instantiation of function template specialization 'JsonValue::read<std::vector<shore_extendable_overmap_terrain_alias, std::allocator<shore_extendable_overmap_terrain_alias> >, nullptr>' requested here
    return ( *member_opt ).read( t, throw_on_error );
                           ^
src/generic_factory.h:828:13: note: in instantiation of function template specialization 'JsonObject::read<std::vector<shore_extendable_overmap_terrain_alias, std::allocator<shore_extendable_overmap_terrain_alias> > >' requested here
    if( !jo.read( name, member ) && !handle_proportional( jo, name, member ) &&
            ^
src/regional_settings.cpp:393:5: note: in instantiation of function template specialization 'optional<std::vector<shore_extendable_overmap_terrain_alias, std::allocator<shore_extendable_overmap_terrain_alias> > >' requested here
    optional( jo, was_loaded, "shore_extendable_overmap_terrain_aliases",
    ^
```
This section describes all the function overloads (non-template functions) it tried to read it with and failed. 
```
src/flexbuffer_json.h:225:14: note: candidate function not viable: no known conversion from 'typename vector<shore_extendable_overmap_terrain_alias, allocator<shore_extendable_overmap_terrain_alias> >::value_type' (aka 'shore_extendable_overmap_terrain_alias') to 'bool &' for 1st argument
        bool read( bool &b, bool throw_on_error = false ) const;
             ^
src/flexbuffer_json.h:226:14: note: candidate function not viable: no known conversion from 'typename vector<shore_extendable_overmap_terrain_alias, allocator<shore_extendable_overmap_terrain_alias> >::value_type' (aka 'shore_extendable_overmap_terrain_alias') to 'char &' for 1st argument
        bool read( char &c, bool throw_on_error = false ) const;
             ^
src/flexbuffer_json.h:227:14: note: candidate function not viable: no known conversion from 'typename vector<shore_extendable_overmap_terrain_alias, allocator<shore_extendable_overmap_terrain_alias> >::value_type' (aka 'shore_extendable_overmap_terrain_alias') to 'signed char &' for 1st argument
        bool read( signed char &c, bool throw_on_error = false ) const;
             ^
src/flexbuffer_json.h:228:14: note: candidate function not viable: no known conversion from 'typename vector<shore_extendable_overmap_terrain_alias, allocator<shore_extendable_overmap_terrain_alias> >::value_type' (aka 'shore_extendable_overmap_terrain_alias') to 'unsigned char &' for 1st argument
        bool read( unsigned char &c, bool throw_on_error = false ) const;
             ^
src/flexbuffer_json.h:229:14: note: candidate function not viable: no known conversion from 'typename vector<shore_extendable_overmap_terrain_alias, allocator<shore_extendable_overmap_terrain_alias> >::value_type' (aka 'shore_extendable_overmap_terrain_alias') to 'unsigned short &' for 1st argument
        bool read( short unsigned int &s, bool throw_on_error = false ) const;
             ^
src/flexbuffer_json.h:230:14: note: candidate function not viable: no known conversion from 'typename vector<shore_extendable_overmap_terrain_alias, allocator<shore_extendable_overmap_terrain_alias> >::value_type' (aka 'shore_extendable_overmap_terrain_alias') to 'short &' for 1st argument
        bool read( short int &s, bool throw_on_error = false ) const;
             ^
src/flexbuffer_json.h:231:14: note: candidate function not viable: no known conversion from 'typename vector<shore_extendable_overmap_terrain_alias, allocator<shore_extendable_overmap_terrain_alias> >::value_type' (aka 'shore_extendable_overmap_terrain_alias') to 'int &' for 1st argument
        bool read( int &i, bool throw_on_error = false ) const;
             ^
src/flexbuffer_json.h:232:14: note: candidate function not viable: no known conversion from 'typename vector<shore_extendable_overmap_terrain_alias, allocator<shore_extendable_overmap_terrain_alias> >::value_type' (aka 'shore_extendable_overmap_terrain_alias') to 'int64_t &' (aka 'long &') for 1st argument
        bool read( int64_t &i, bool throw_on_error = false ) const;
             ^
src/flexbuffer_json.h:233:14: note: candidate function not viable: no known conversion from 'typename vector<shore_extendable_overmap_terrain_alias, allocator<shore_extendable_overmap_terrain_alias> >::value_type' (aka 'shore_extendable_overmap_terrain_alias') to 'uint64_t &' (aka 'unsigned long &') for 1st argument
        bool read( uint64_t &i, bool throw_on_error = false ) const;
             ^
src/flexbuffer_json.h:234:14: note: candidate function not viable: no known conversion from 'typename vector<shore_extendable_overmap_terrain_alias, allocator<shore_extendable_overmap_terrain_alias> >::value_type' (aka 'shore_extendable_overmap_terrain_alias') to 'unsigned int &' for 1st argument
        bool read( unsigned int &u, bool throw_on_error = false ) const;
             ^
src/flexbuffer_json.h:235:14: note: candidate function not viable: no known conversion from 'typename vector<shore_extendable_overmap_terrain_alias, allocator<shore_extendable_overmap_terrain_alias> >::value_type' (aka 'shore_extendable_overmap_terrain_alias') to 'float &' for 1st argument
        bool read( float &f, bool throw_on_error = false ) const;
             ^
src/flexbuffer_json.h:236:14: note: candidate function not viable: no known conversion from 'typename vector<shore_extendable_overmap_terrain_alias, allocator<shore_extendable_overmap_terrain_alias> >::value_type' (aka 'shore_extendable_overmap_terrain_alias') to 'double &' for 1st argument
        bool read( double &d, bool throw_on_error = false ) const;
             ^
src/flexbuffer_json.h:237:14: note: candidate function not viable: no known conversion from 'typename vector<shore_extendable_overmap_terrain_alias, allocator<shore_extendable_overmap_terrain_alias> >::value_type' (aka 'shore_extendable_overmap_terrain_alias') to 'std::string &' (aka 'basic_string<char> &') for 1st argument
        bool read( std::string &s, bool throw_on_error = false ) const;
             ^
```

And this is the complicated part, where it tries to match templates, and checks if it can read the type based on certain characteristics it has.
```
src/flexbuffer_json.h:271:14: note: candidate template ignored: could not match 'array<type-parameter-0-0, _Nm>' against 'shore_extendable_overmap_terrain_alias'
        bool read( std::array<T, N> &v, bool throw_on_error = false ) const;
             ^
src/flexbuffer_json-inl.h:275:17: note: candidate template ignored: could not match 'string_id<type-parameter-0-0>' against 'shore_extendable_overmap_terrain_alias'
auto JsonValue::read( string_id<T> &thing, bool throw_on_error ) const -> bool
                ^
src/flexbuffer_json-inl.h:287:17: note: candidate template ignored: could not match 'int_id<type-parameter-0-0>' against 'shore_extendable_overmap_terrain_alias'
auto JsonValue::read( int_id<T> &thing, bool throw_on_error ) const -> bool
                ^
src/flexbuffer_json.h:278:14: note: candidate template ignored: substitution failure [with T = shore_extendable_overmap_terrain_alias]: no type named 'key_type' in 'shore_extendable_overmap_terrain_alias'
        bool read( T &v, bool throw_on_error = false ) const;
             ^
src/flexbuffer_json-inl.h:299:17: note: candidate template ignored: substitution failure [with T = shore_extendable_overmap_terrain_alias]: no matching function for call to 'deserialize'
auto JsonValue::read( T &v, bool throw_on_error ) const ->
                ^
src/flexbuffer_json-inl.h:315:17: note: candidate template ignored: substitution failure [with T = shore_extendable_overmap_terrain_alias]: no member named 'deserialize' in 'shore_extendable_overmap_terrain_alias'
auto JsonValue::read( T &v, bool throw_on_error ) const -> decltype( v.deserialize( *this ),
                ^                                                      ~~~~~~~~~~~
src/flexbuffer_json-inl.h:330:17: note: candidate template ignored: requirement 'std::is_enum_v<shore_extendable_overmap_terrain_alias>' was not satisfied [with T = shore_extendable_overmap_terrain_alias]
bool JsonValue::read( T &val, bool throw_on_error ) const
                ^
src/flexbuffer_json.h:282:14: note: candidate template ignored: could not match 'enum_bitset<type-parameter-0-0>' against 'shore_extendable_overmap_terrain_alias'
        bool read( enum_bitset<T> &v, bool throw_on_error = false ) const;
             ^
src/flexbuffer_json.h:287:14: note: candidate template ignored: could not match 'colony<type-parameter-0-0, allocator<type-parameter-0-0>, unsigned short>' against 'shore_extendable_overmap_terrain_alias'
        bool read( cata::colony<T> &v, bool throw_on_error = false ) const;
             ^
src/flexbuffer_json.h:293:14: note: candidate template ignored: could not match 'colony<type-parameter-0-0, allocator<type-parameter-0-0>, unsigned short>' against 'shore_extendable_overmap_terrain_alias'
        bool read( cata::colony<T> &v, bool throw_on_error = false ) const;
             ^
src/flexbuffer_json-inl.h:347:17: note: candidate template ignored: could not match 'pair<type-parameter-0-0, type-parameter-0-1>' against 'shore_extendable_overmap_terrain_alias'
bool JsonValue::read( std::pair<T, U> &p, bool throw_on_error ) const
                ^
src/flexbuffer_json.h:300:14: note: candidate template ignored: substitution failure [with T = shore_extendable_overmap_terrain_alias]: no type named 'key_type' in 'shore_extendable_overmap_terrain_alias'
        bool read( T &m, bool throw_on_error = true ) const;
             ^
src/flexbuffer_json-inl.h:374:17: note: candidate template ignored: substitution failure [with T = shore_extendable_overmap_terrain_alias]: no type named 'value_type' in 'shore_extendable_overmap_terrain_alias'
auto JsonValue::read( T &v, bool throw_on_error ) const -> decltype( v.front(), true )
                ^
```
Now that's a lot of noise, but if you zoom in on these two
```
src/flexbuffer_json-inl.h:299:17: note: candidate template ignored: substitution failure [with T = shore_extendable_overmap_terrain_alias]: no matching function for call to 'deserialize'
auto JsonValue::read( T &v, bool throw_on_error ) const ->
                ^
src/flexbuffer_json-inl.h:315:17: note: candidate template ignored: substitution failure [with T = shore_extendable_overmap_terrain_alias]: no member named 'deserialize' in 'shore_extendable_overmap_terrain_alias'
auto JsonValue::read( T &v, bool throw_on_error ) const -> decltype( v.deserialize( *this ),
                ^                                                      ~~~~~~~~~~~
```
These ones are looking for either a `deserialize( shore_extendable_overmap_terrain_alias &, JsonValue & )` or `shore_extendable_overmap_terrain_alias::deserialize( JsonValue & )` function. So if one of those functions are defined, it will now be able to read them with that previous method.

https://github.com/CleverRaven/Cataclysm-DDA/blob/e3691d2d6900ffc7e7d0dc4b2b948778cf6c6c83/src/flexbuffer_json-inl.h#L297-L328

If this is all gobbledygook, that's fine - it seems to me that stuff like this is one of the least understood aspects of C++. The valuable takeaway from all of this is that if you add a `deserialize` function to a type, you can easily read it without having to do all the stuff manually like this does.
</details>

Reply:
thanks, and thanks for the breakdown of why it helps.